docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+    information can be omitted from the StorageClass
+- The *volume_id* carries information regarding the *cluster_id* and pool,
+  enabling *volume_id* based RPC requests to operate using the provided
+  configuration information against the respective cluster
+- Usual CSI spec based maps such as, *volume_context* and *publish_context*
+  continue to carry information regarding required aspects about the
+  provisioned volume to dependent RPCs
+- The NS also has access to the config maps and secrets to satisfy required node
+  level RPC requests
+
+### Pros
+
+- Addresses removal of the current config map that maintains list of
+  provisioned volumes
+- Provides the ability for the CSI plugin to support all RPCs, as required
+  information is present with the plugin across all Ceph clusters
+- Safeguards secrets and cluster configuration information access to only
+  required namespaces that need to manage and use it
+- Provides the ability to have a single Ceph-CSI plugin instance per CO
+  instance, even when the CO instance needs volumes provisioned from multiple
+  Ceph clusters
+- Also, does not take away the capability, where needed, to use multiple
+  Ceph-CSI plugin instances (specialized using provisioner name) per CO
+  instance
+- Leaves the flexibility for the StorageClass to still override secrets, if such
+  requirements arise
+  - **NOTE:** The credentials stored for a cluster and passed to the CSI
+    plugins CS and NS, should be able to list volumes/images created using
+    other secrets (IOW, should act like a supervisor/root) to correctly
+    support ListVolumes and ListSnapshots operations
+- Provides the ability to deal with multi-cluster deployments, and possible
+  best-fit provisioning in the **future**
+  - Enables the ability for the CSI plugin to act on Topology based requests
+  - Simplifies StorageClass definitions, leaving the choice of the
+    cluster/pool to the CSI plugin
+
+### Cons
+
+- Addition/Removal of Ceph cluster configuration, to the CSI plugin, requires
+  that the pod specification be updated, and thus the CS and NS pods require
+  to be restarted, in order to pick up the latest cluster details
+  - If done right, the NS restarts should not impact existing volumes that
+    are staged on the node, and the CS should have multiple instances
+    running thus enabling restarting instances one at a time without
+    impacting the provisioning service
+
+### Configuration specifics
+
+#### `ceph-cluster-<cluster-fsid>` config map contents JSON representation
+
+```
+{
+  "<clusterID>": {
+    "monitors": [
+      "IP/DNS:port",
+      ...
+    ],
+    "pools": [
+      <pool-name>,
+      ...
+    ]
+  }
+}
+```
+
+#### `ceph-cluster-<cluster-fsid>-[provisioner|publish]-secret` definition
+
+```
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ceph-cluster-<cluster-fsid>-*-secret
+  namespace: default
+data:
+  # ID of the admin/user
+  subjectid: BASE64-ENCODED-ID
```

Comment:
I believe I already saw a conversation about the correct terminology for ceph credentials...can't remember where exactly. Anyway, `subjectid` and `credentials` seem a bit confusing. Credentials usually refer to a username-password pair - as opposed to just the "password" part. AFAIK passwords are referred to as keys in ceph. As for the username part, according to the [documentation](http://docs.ceph.com/docs/master/rados/operations/user-management/), and the way the names are constructed (`TYPE.ID` - in our case the `TYPE` will always be `client`), how about using `clientid`?

`clientid` in the role of username, `key` for password. Thoughts?

Reply:
This has been renamed as adminid/user/id and adminkey/userkey as per other comments that note the same, see [here](https://github.com/ceph/ceph-csi/pull/241/files#diff-7551e1d8a03243f545e81cfd4ecd9156R25). Let me know if this works.